File: python/54.py
# stdlib
from time import time
from collections import Counter


def preprocess_hand(line):
    line = line.replace("T", "10").replace("J", "11").replace("Q", "12").replace("K", "13").replace("A", "14")
    cards = [(int(card[:-1]), card[-1]) for card in line.split()]
    return cards[:5], cards[5:]

#################
# Hand Rankings #
#################


def get_rank(cards):
    sorted_values = sorted([x[0] for x in cards])
    counter = Counter([x[0] for x in cards])
    if len(set([x[1] for x in cards])) == 1 and not {"10", "11", "12", "13", "14"} - set([x[0] for x in cards]):
        # royal flush
        return 10, sorted_values[-1], sorted_values
    elif len(set([x[1] for x in cards])) == 1 and sorted_values == list(range(sorted_values[0], sorted_values[-1] + 1)):
        # straight_flush
        return 9, sorted_values[-1], sorted_values
    elif counter.most_common(1)[0][1] == 4:
        # four of a kind
        return 8, counter.most_common(1)[0][0], sorted_values
    elif counter.most_common(1)[0][1] == 3 and any([x[1] == 2 for x in counter.most_common(2)]):
        # full house
        return 7, counter.most_common(1)[0][0], sorted_values
    elif len(set([x[1] for x in cards])) == 1:
        # flush
        return 6, sorted_values[-1], sorted_values
    elif sorted_values == list(range(sorted_values[0], sorted_values[-1] + 1)):
        # straight
        return 5, sorted_values[-1], sorted_values
    elif counter.most_common(1)[0][1] == 3:
        # three of a kind
        return 4, counter.most_common(1)[0][0], sorted_values
    elif all([x[1] == 2 for x in counter.most_common(2)]):
        # two pairs
        return 3, max([x[0] for x in counter.most_common(2)]), sorted_values
    elif any([x[1] == 2 for x in counter.most_common(2)]):
        # one pair
        return 2, counter.most_common(1)[0][0], sorted_values
    else:
        # high card
        return 1, max([x[0] for x in cards]), sorted_values

###################
# ranking players #
###################


def is_player_1_winning(game):
    player1, player2 = game
    rank1, max_rank1, sorted_values1 = get_rank(player1)
    rank2, max_rank2, sorted_values2 = get_rank(player2)
    if rank1 != rank2:
        return rank1 > rank2
    elif max_rank1 != max_rank2:
        return max_rank1 > max_rank2
    for i in range(5):
        if sorted_values1[-1 - i] != sorted_values2[-1 - i]:
            return sorted_values1[-1 - i] > sorted_values2[-1 - i]
    return True


if __name__ == "__main__":
    since = time()
    games = [preprocess_hand(l.strip()) for l in open("p054_poker.txt").readlines()]

    count = sum([is_player_1_winning(game) for game in games])
    print(f"Player 1 won {count} times!")
    print(f"Took {round(time() - since, 2)} s")

File: cpp/54.cpp
#include <iostream>
#include <string>
#include <vector>
#include <fstream>
#include <algorithm>
#include <boost/algorithm/string/replace.hpp>
#include <sstream>
#include <map>
#include <set>
#include <numeric>

using namespace std;
using namespace boost;

vector<vector<string>> preprocess_hand(string line)
{
    replace_all(line, "T", "10");
    replace_all(line, "J", "11");
    replace_all(line, "Q", "12");
    replace_all(line, "K", "13");
    replace_all(line, "A", "14");
    vector<vector<string>> cards{};
    string token;
    istringstream tokenStream(line);
    while (getline(tokenStream, token, ' '))
    {
        cards.push_back({token.substr(0, token.length() - 1), token.substr(token.length() - 1, 1)});
    }
    return cards;
}

vector<vector<int>> counter(vector<int> values)
{
    vector<vector<int>> count{};

    map<int, int> occurrences{};
    for (int i : values)
    {
        if (occurrences.find(i) != occurrences.end())
            occurrences[i] += 1;
        else
            occurrences[i] = 1;
    }
    for (auto [key, value] : occurrences)
    {
        count.push_back({key, value});
    }
    sort(count.begin(), count.end(),
         [](const std::vector<int> &a, const std::vector<int> &b) {
             return a[1] > b[1];
         });
    return count;
}

///////////////////
// Hand Rankings //
///////////////////

vector<vector<int>> get_rank(vector<vector<string>> cards)
{
    vector<vector<int>> result{};
    vector<int> sorted_values{};
    set<string> symb{};
    for (int i = 0; i < cards.size(); i++)
    {
        sorted_values.push_back(stoi(cards[i][0]));
        symb.insert(cards[i][0]);
    }
    vector<vector<int>> counter_ = counter(sorted_values);
    sort(sorted_values.begin(), sorted_values.end());

    // royal flush
    vector<int> royal{10, 11, 12, 13, 14};
    if (symb.size() == 1 && sorted_values == royal)
    {
        result = {{10}, {sorted_values.back()}, sorted_values};
        return result;
    }

    // straight flush
    vector<int> straight(5);
    iota(straight.begin(), straight.end(), sorted_values[0]);
    if (symb.size() == 1 && straight == sorted_values)
    {
        result = {{9}, {sorted_values.back()}, sorted_values};
        return result;
    }

    // four of a kind
    if (counter_[0][1] == 4)
    {
        result = {{8}, {counter_[0][0]}, sorted_values};
        return result;
    }

    // full house
    if (counter_[0][1] == 3 && counter_[1][1] == 2)
    {
        result = {{7}, {counter_[0][0]}, sorted_values};
        return result;
    }

    // flush
    if (symb.size() == 1)
    {
        result = {{6}, {sorted_values.back()}, sorted_values};
        return result;
    }

    // straight
    if (straight == sorted_values)
    {
        result = {{5}, {sorted_values.back()}, sorted_values};
        return result;
    }

    // three of a kind
    if (counter_[0][1] == 3)
    {
        result = {{4}, {counter_[0][0]}, sorted_values};
        return result;
    }

    //  two pair
    if (counter_[0][1] == 2 and counter_[1][1] == 2)
    {
        result = {{3}, {max(counter_[0][0], counter_[1][0])}, sorted_values};
        return result;
    }

    // one pair
    if (counter_[0][1] == 2)
    {
        result = {{2}, {counter_[0][0]}, sorted_values};
        return result;
    }

    // high card
    result = {{1}, {sorted_values.back()}, sorted_values};
    return result;
}

//////////////////////
// Rankings Players //
/////////////////////

bool is_player_1_winning(vector<vector<string>> game)
{
    vector<vector<string>> player1(game.begin(), game.begin() + 5);
    vector<vector<string>> player2(game.begin() + 5, game.begin() + 10);
    vector<vector<int>> result1 = get_rank(player1);
    vector<vector<int>> result2 = get_rank(player2);
    if (result1[0][0] != result2[0][0])
        return result1[0][0] > result2[0][0];
    else if (result1[1][0] != result2[1][0])
        return result1[1][0] > result2[1][0];
    for (int i = 0; i < 5; i++)
    {
        if (result1[2][4 - i] != result2[2][4 - i])
            return result1[2][4 - i] > result2[2][4 - i];
    }
    return true;
}

int main()
{
    int count{};
    fstream file("54.txt");
    string line{};
    vector<string> hands{};
    vector<vector<string>> cards{};

    while (getline(file, line))
    {
        hands.push_back(line);
    }

    for (string hand : hands)
    {
        cards = preprocess_hand(hand);
        if (is_player_1_winning(cards))
            count++;
    }

    cout << count << endl;
    return 0;
}